### PR TITLE
[SPARK-40212] [SQL] SparkSQL castPartValue does not properly handle byte, short, or float

### PR DESCRIPTION
**Pull request by BrennanStein**
### What changes were proposed in this pull request?
The `castPartValueToDesiredType` function now returns byte for ByteType and short for ShortType, rather than ints; also floats for FloatType rather than double.

### Why are the changes needed?
Previously, attempting to read back in a file partitioned on one of these column types would result in a ClassCastException at runtime (for Byte, `java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Byte`). I can't think this is anything but a bug, as returning the correct data type prevents the crash.

### Does this PR introduce _any_ user-facing change?
Yes: it changes the observed behavior when reading in a byte/short/float-partitioned file.

### How was this patch tested?
Added unit test. Without the `castPartValueToDesiredType` updates, the test fails with the stated exception.

===
I'll note that I'm not familiar enough with the spark repo to know if this will have ripple effects elsewhere, but tests pass on my fork and since the very similar https://github.com/apache/spark/pull/36344/files only needed to touch these two files I expect this change is self-contained as well.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 1]
```diff
@@ -530,7 +530,9 @@ object PartitioningUtils extends SQLConfHelper {
     case _ if value == DEFAULT_PARTITION_NAME => null
     case NullType => null
     case StringType => UTF8String.fromString(unescapePathName(value))
-    case ByteType | ShortType | IntegerType => Integer.parseInt(value)
+    case ByteType => Integer.parseInt(value).toByte
+    case ShortType => Integer.parseInt(value).toShort
+    case IntegerType => Integer.parseInt(value)
     case LongType => JLong.parseLong(value)
     case FloatType | DoubleType => JDouble.parseDouble(value)
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:537]: It seems `FloatType` also has this issue.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -4371,6 +4371,23 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     }
   }
 
+  test("SPARK-40212: SparkSQL castPartValue does not properly handle byte & short") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4374]: Move the test to `ParquetPartitionDiscoverySuite`?

Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Please also add this fix to PR title and PR description.

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
cc @cloud-fan @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by cloud-fan — COMMENTED** [review]
good catch! late LGTM

**Pull request closed without merging**